accepted/cranelift-backend-transition.md

```diff
@@ -0,0 +1,215 @@
+[#1936](bytecodealliance/wasmtime#1936); this RFC extends @bnjbvr's
+initial proposal and excellent fact-finding work in that issue in
+order to get final sign-off and finish the transition.
+
+# Proposal
+[proposal]: #proposal
+
+At a high level, we propose (i) evaluating criteria by which we decide
+to make the transition; (ii) obtaining the appropriate sign-offs; and
+(iii) transitioning in several steps in a way that is designed to
+preserve stability and leave options if anything goes wrong.
+
+## Transition Criteria
+
+1. *Feature-completeness*: All needed functionality must be present and
+  all tests must pass. 
+  
+  -  Status: we are largely there already. In
+     bytecodealliance/wasmtime#2718, a trial/draft PR, the default was
+     switched and all tests that are still applicable (i.e., not testing
+     specific details of the old backend) are passing. Only
+     bytecodealliance/wasmtime#2710 needs to land for the new backend to
+     be fully feature-complete.
+   
+     - Note: the new backend does not fully support Wasm-SIMD. However,
+       the old backend no longer does, either, after the proposal's
+       recent evolution. Hence, we are willing to accept this
+       incompleteness because it does not a regression overall.
+   
+2. *Performance*: The new backend has acceptable compilation speed and
+  generates acceptably good code.
+  
+   - Earlier evaluations showed that we were trending this way. No
+     recent comprehensive comparison has been done, however. This RFC
+     proposes to use the
+     [Sightglass](https://github.com/bytecodealliance/sightglass/)
+     benchmark suite in order to evaluate both dimensions.
+   
+   - Compilation time: we should expect as good or better compilation
+     time on most benchmarks. We have seen some cases where the register
+     allocator in the new backend can be slower on very large
+     inputs. This RFC proposes to balance such slowdowns against the
+     other benefits of the transition and to cautiously accept a small
+     budget for some such cases, in tandem with a priority effort to
+     understand the slowdown and address it soon.
+   
+     - Open question: how much degradation should we accept?
+   
+   - Runtime: we should expect as good or better code generated on most
+     benchmarks.
+   
+     - Open question: how much degradation should we accept?
+
+3. *Compatibility*: there should be no known or open blocking issues
+   with the new backend when used by any significant user; in other
+   words, we should not break anyone, and if the transition would do
+   so, we should work with these stakeholder(s) first to work around
+   the issue.
+  
+   - Status: cg\_clif and Lucet have already transitioned to the new
+     backend. Wasmtime is compatible. Firefox uses Cranelift only on
+     aarch64, which is already using the new backend framework.
+
+     - Open question: are there other projects with which we should
+       consult?
```

Comment:
Maybe saltwater by @jyn514? It has been a couple of months since the last change through it seems.

Reply:
I don't really work much on it any more, it uses like a 6 month old version on cranelift. I wouldn't make decisions based on it.